### DISCUSSION TITLE
caught-error / "Disk space error"

### DISCUSSION THREAD

**Discussion by janpio** [renovatebot/renovate#8460] [Request Help]
We are _super_ happy users of Renovate, but recently hit a small snag:

**What Renovate type, platform and version are you using?**

Hosted GitHub App

**Describe the bug**

One of our repositories https://github.com/prisma/e2e-tests stopped running Renovate successfully. 
From https://app.renovatebot.com/ we know that all jobs report `caught-error` as result. 

**Relevant debug logs**

In the details of a job we can see the following:

```
...

DEBUG: lock file error(branch="renovate/node-12.x")
{
  "err": {
    "killed": false,
    "code": 1,
    "signal": null,
    "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/prisma/e2e-tests\":\"/mnt/renovate/gh/prisma/e2e-tests\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/home/ubuntu/.npmrc\":\"/home/ubuntu/.npmrc\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs\" docker.io/renovate/node:14.15.4 bash -l -c \"npm i -g yarn && yarn install\"",
    "stdout": "\n> yarn@1.22.10 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.10\nadded 1 package in 1.43s\nyarn install v1.22.10\n[1/4] Resolving packages...\n[2/4] Fetching packages...\ninfo fsevents@2.1.2: The platform \"linux\" is incompatible with this module.\ninfo \"fsevents@2.1.2\" is an optional dependency and failed compatibility check. Excluding it from installation.\ninfo fsevents@1.2.12: The platform \"linux\" is incompatible with this module.\ninfo \"fsevents@1.2.12\" is an optional dependency and failed compatibility check. Excluding it from installation.\n[3/4] Linking dependencies...\ninfo If you think this is a bug, please open a bug report with the information provided in \"/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs/yarn-error.log\".\ninfo Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.\n",
    "stderr": "warning ../../package.json: No license field\nerror An unexpected error occurred: \"ENOSPC: no space left on device, copyfile '/home/ubuntu/.cache/yarn/v6/npm-typescript-4.1.3-519d582bd94cba0cf8934c7d8e8467e473f53bb7-integrity/node_modules/typescript/lib/typescriptServices.js' -> '/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs/node_modules/typescript/lib/typescriptServices.js'\".\n",
    "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/prisma/e2e-tests\":\"/mnt/renovate/gh/prisma/e2e-tests\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/home/ubuntu/.npmrc\":\"/home/ubuntu/.npmrc\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs\" docker.io/renovate/node:14.15.4 bash -l -c \"npm i -g yarn && yarn install\"\nwarning ../../package.json: No license field\nerror An unexpected error occurred: \"ENOSPC: no space left on device, copyfile '/home/ubuntu/.cache/yarn/v6/npm-typescript-4.1.3-519d582bd94cba0cf8934c7d8e8467e473f53bb7-integrity/node_modules/typescript/lib/typescriptServices.js' -> '/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs/node_modules/typescript/lib/typescriptServices.js'\".\n",
    "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/prisma/e2e-tests\":\"/mnt/renovate/gh/prisma/e2e-tests\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/home/ubuntu/.npmrc\":\"/home/ubuntu/.npmrc\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs\" docker.io/renovate/node:14.15.4 bash -l -c \"npm i -g yarn && yarn install\"\nwarning ../../package.json: No license field\nerror An unexpected error occurred: \"ENOSPC: no space left on device, copyfile '/home/ubuntu/.cache/yarn/v6/npm-typescript-4.1.3-519d582bd94cba0cf8934c7d8e8467e473f53bb7-integrity/node_modules/typescript/lib/typescriptServices.js' -> '/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs/node_modules/typescript/lib/typescriptServices.js'\".\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:314:20)\n    at ChildProcess.EventEmitter.emit (domain.js:483:12)\n    at maybeClose (internal/child_process.js:1021:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)"
  },
  "type": "yarn"
}
DEBUG: Passing disk-space error up(branch="renovate/node-12.x")
ERROR: Disk space error - skipping
...
```
(Taken from https://app.renovatebot.com/dashboard#github/prisma/e2e-tests/293576384)

All our other repositories are working fine.

Is there something we are doing wrong or is this "just" an infrastructure problem?

**Comment by HonkingGoose** [comment]
If the other repo's don't have this problem, with the same config, and this repo used to work properly but now doesn't, one logical explanation could be that the repo filesize grew too big.

How big is your repository when you do a standard full-fat `git clone`?
It might be that you're running into file-size restrictions for the Docker container...

There's an issue open to support doing a partial clone for big repositories: https://github.com/renovatebot/renovate/issues/7840

**Reply by janpio** [reply]
A `git clone` creates a ~220MB folder on my disk. Doesn't sound like a size where limitations would already be hit. Right?
The repo is unusual in that is contains multiple, around 50, projects that have a `package.json` file.

**Reply by HonkingGoose** [reply]
> A `git clone` creates a ~220MB folder on my disk. Doesn't sound like a size where limitations would already be hit. Right?

I agree, that shouldn't be a size where you hit any reasonable limit, I guess.

> The repo is unusual in that is contains multiple, around 50, projects that have a `package.json` file.

Renovate should handle mono-repos with a lot of package.json files just fine.
But don't quote me on that, as I'm not the head maintainer... 😄 

---

How big does the repo get once you install all the dependencies?
It seems to error out when copying a file, so maybe it's hitting the file-size limit during that operation?

```
An unexpected error occurred: 
"ENOSPC: no space left on device, copyfile '/home/ubuntu/.cache/yarn/v6/npm-typescript-4.1.3-519d582bd94cba0cf8934c7d8e8467e473f53bb7-integrity/node_modules/typescript/lib/typescriptServices.js' -> '/mnt/renovate/gh/prisma/e2e-tests/frameworks/nextjs/node_modules/typescript/lib/typescriptServices.js'\".\n",
```

**Reply by rarkins** [reply]
Seems like it might hit the limit once all node modules are installed. I can change this repo from SSD to EBS disk where there's more space

**Reply by rarkins** [reply]
It seems to be running fine now, with the increased disk space

**Reply by janpio** [reply]
Thank you @rarkins, can confirm everything is working fine now.

**Reply by janpio** [reply]
Ohoh, seems we are getting similar errors again @rarkins:
```
...
 "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/gh/prisma/e2e-tests\":\"/tmp/renovate/gh/prisma/e2e-tests\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/home/ubuntu/.npmrc\":\"/home/ubuntu/.npmrc\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/gh/prisma/e2e-tests/frameworks/nestjs\" docker.io/renovate/node:14.16.0 bash -l -c \"npm i -g yarn && yarn install\"\nwarning ../../package.json: No license field\nwarning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.\nerror An unexpected error occurred: \"https://registry.npmjs.org/typescript/-/typescript-4.2.3.tgz: ENOSPC: no space left on device, write\".\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:315:20)\n    at ChildProcess.EventEmitter.emit (domain.js:467:12)\n    at maybeClose (internal/child_process.js:1048:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
  },
  "type": "yarn"
}
DEBUG: Passing disk-space error up(branch="renovate/typescript-4.x")
ERROR: Disk space error - skipping
```
https://app.renovatebot.com/dashboard#github/prisma/e2e-tests/319604116

**Reply by rarkins** [reply]
Sorry but we have no further disk space to allocate. We don't allocate worker machines per-account or per-repo so to increase it just for your case would mean increasing space and costs of all 150 workers. If you have really large repos then running self-hosted (and free) WhiteSource Renovate On-Premises may be a better option.

**Reply by janpio** [reply]
That makes a lot of sense. I requested some information on that and will see what needs to be done there.

Do you have an intuitive feeling _what_ is using that much space? Is there anything we can do to maybe keep this working for a bit? Can we somehow configure something to do less things? (I know that our postinstall hook for example downloads binaries, ~50MB per `package.json` which we have ~50 in that project)

**Reply by rarkins** [reply]
FYI some of the machines have been up a while and we don't have any automatic Docker image vacuuming yet, so you may have got unlucky and a machine with less space than others. I terminated/recreated all older ones but it's not a guarantee. Does it fail every run?

50x50MB seems pretty heavy if it can be avoided. Is it the same 50MB which could somehow be sym linked?

**Reply by Jolg42** [reply]
It should be the same 50MB of binaries yes, so something like pnpm should be able to symlink
Or maybe Renovate can run with the equivalent of `npm install --ignore-scripts`?

**Reply by janpio** [reply]
Unfortunately not - but there is a yarn (1) argument to skip these steps (`--ignore-scripts`) and an env var to disable the code on our side - but I am not sure we can/could set these for the Renovate Cloud app.

**Reply by Jolg42** [reply]
The failure started 3 hours ago so maybe the termination will help for next run
<img width="1056" alt="Screen Shot 2021-03-18 at 14 53 53" src="https://user-images.githubusercontent.com/1328733/111637516-c95f3b00-87f9-11eb-9b9d-2d0ad6863320.png">

**Reply by rarkins** [reply]
In general Renovate does try to ignore scripts. 

It's strange you had 3 errors in a row, as they would likely have been on 3 separate machines. Unless all our machines suddenly crossed some threshold necessary to run your repo, it would seem like the failures are related either to changes in your or our source code.

**Reply by janpio** [reply]
That's also what confuses me: We did not change anything on the repo but update a few dependencies ourselves, and a bunch of Renovate updates as well: https://github.com/prisma/e2e-tests/commits/dev

**Reply by Jolg42** [reply]
I removed some dependencies, not sure if that did the trick or your restart but it's back to "Done" which means it works.
<img width="1005" alt="Screen Shot 2021-03-18 at 17 58 03" src="https://user-images.githubusercontent.com/1328733/111666943-7f378300-8814-11eb-908c-99845a69378c.png">
